src/smt/store/smt_store.rs

```diff
@@ -0,0 +1,163 @@
+use crate::smt::store::serde::leaf_key_to_vec;
+use crate::smt::types::leaf::{Byte32, SMTLeaf, SMTLeafBuilder, SMTLeafVec, SMTLeafVecBuilder};
+use crate::utils::error::Error;
+use crate::utils::helper::parse_vec_n;
+use cota_smt::smt::H256;
+use molecule::prelude::{Builder, Entity};
+use sparse_merkle_tree::{
+    error::Error as SMTError,
+    traits::Store,
+    tree::{BranchKey, BranchNode},
+};
+
+pub struct SMTStore<'a> {
+    lock_hash:  [u8; 32],
+    leaf_col:   Col,
+    branch_col: Col,
+    root_col:   Col,
+    leaves_col: Col,
+    store:      &'a CotaRocksDB,
+}
+
+impl<'a> SMTStore<'a> {
+    pub fn new(
+        lock_hash: [u8; 32],
+        leaf_col: Col,
+        branch_col: Col,
+        root_col: Col,
+        leaves_col: Col,
+        store: &'a CotaRocksDB,
+    ) -> Self {
+        SMTStore {
+            lock_hash,
+            leaf_col,
+            branch_col,
+            root_col,
+            leaves_col,
+            store,
+        }
+    }
+
+    pub fn save_root(&self, root: &H256) -> Result<(), SMTError> {
+        self.store
+            .insert_raw(self.root_col, &self.lock_hash, root.as_slice())
+            .map_err(|err| SMTError::Store(format!("insert error {:?}", err)))?;
+        Ok(())
+    }
+
+    pub fn get_root(&self) -> Result<Option<H256>, SMTError> {
+        match self.store.get(self.root_col, &self.lock_hash) {
+            Some(slice) => Ok(Some(H256::from(parse_vec_n(slice.to_vec())))),
+            None => Ok(None),
+        }
+    }
+
+    pub fn insert_leaves(&self, leaves: Vec<(H256, H256)>) -> Result<(), Error> {
+        let smt_leaf_vec: Vec<SMTLeaf> = leaves
+            .into_iter()
+            .map(|leaf| {
+                SMTLeafBuilder::default()
+                    .key(Byte32::from_slice(leaf.0.as_slice()).unwrap())
+                    .value(Byte32::from_slice(leaf.1.as_slice()).unwrap())
+                    .build()
+            })
+            .collect();
+        let smt_leaves: SMTLeafVec = SMTLeafVecBuilder::default().set(smt_leaf_vec).build();
+        self.store
+            .insert_raw(self.leaves_col, &self.lock_hash, smt_leaves.as_slice())
+            .map_err(|err| Error::SMTError(format!("insert error {:?}", err)))?;
+
+        Ok(())
+    }
+
+    pub fn get_leaves(&self) -> Result<Option<Vec<(H256, H256)>>, Error> {
+        match self.store.get(self.leaves_col, &self.lock_hash) {
+            Some(slice) => {
+                let mut leaves = vec![];
+                let smt_leaves = SMTLeafVec::from_slice(&slice)
+                    .map_err(|_e| Error::SMTError("SMT Leaves parse error".to_owned()))?;
+                for smt_leaf in smt_leaves {
+                    leaves.push((
+                        H256::from(parse_vec_n::<32>(smt_leaf.key().as_slice().to_vec())),
+                        H256::from(parse_vec_n::<32>(smt_leaf.value().as_slice().to_vec())),
+                    ))
+                }
+                Ok(Some(leaves))
```

Comment:
same as previous suggestion, replace `parse_vec_n` with array conversion and replace vec allocation / push with iter map / collect:

```suggestion
                let smt_leaves = SMTLeafVec::from_slice(&slice)
                    .map_err(|_e| Error::SMTError("SMT Leaves parse error".to_owned()))?;
                let leaves = smt_leaves.into_iter().map(|smt_leaf| {
                    let key: [u8; 32] = smt_leaf.key().as_slice().try_into().expect("stored SMTLeaf should be valid");
                    let value: [u8; 32] = smt_leaf.value().as_slice().try_into().expect("stored SMTLeaf should be valid");
                    (
                        key.into(),
                        value.into(),
                    )
                }).collect();
                Ok(Some(leaves))
```